text/0000-cargo-asymmetric-tokens.md

```diff
@@ -0,0 +1,218 @@
+After the asymmetric token has expired, the data sent over the network can be made public, without risking the private material. A registry can keep or publish an audit log of asymmetric tokens without risk of them being reused, in case a security auditor would like to look for abnormal or unusual behavior.
+
+Different registries will have different users in mind and have different use cases. Therefore, they will need to have different behaviors. So, there are many decisions a registry has to make that this RFC has no opinion on. Some examples:
+
+- Bootstrapping trust: how does the registry decide to trust a new user?
+- Key generation: where and how is the key pair made?
+- Key rotation: how often do existing users need to make a new key pair?
+- Revocation: how does the registry decide to stop trusting an existing key pair?
+
+In order for crates.io to support asymmetric tokens these questions will need to be answered for crates.io. If and how crates.io will implement compatibility with these new tokens will be left for a follow-up discussion/RFC.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Private registries that require authentication use asymmetric cryptography as a more secure way for cargo to log in. Each registry works a little different, but the most common workflow is:
+1. Generate a key pair. (For many registries, you can generate the key pair using the cargo command `cargo login --registry=name --generate-keypair`, which will print the public key for use in the next step.)
+2. Log into the registries website
+3. Go to the "register a key pair" page, upload your public key. and get the user ID for that key pair.
+
+Most do not, but some registries require one more step:
+4. if the registry gave you a `key-subject` then on the command line run `cargo login --registry=name --key-subject="the provided data"`
+
+There are credential processes for using key pairs stored on hardware tokens. Check crates.io to see if there's one available for your hardware. Each one is a little different, but the general workflow is:
+1. `cargo install credential-process-for-your-hardware-token`
+2. Run `cargo credential-process-for-your-hardware-token setup registryURL` to get your public key.
+3. Edit `credentials.toml` to have a `credential-process` field as described by `credential-process-for-your-hardware-token` docs. (The credential process command may help do this for you.)
+4. Log into the registries website
+5. Go to the "register a key pair" page, upload your public key.
+
+Some registries prioritize user experience over strictest security. They can simplify the process by providing key generation on the server. If your registry works this way the workflow will be:
+1. Log into the registries website
+2. Go to the "register generate a key pair" page, and copy the command it generated for you. It will disappear when you leave the page, the server will not keep a copy of the public key!
```

Comment:
Should this be "the server will not keep a copy of the **private** key"?

Reply:
Thank you for the correction.